File: vscode/diagram-renderer/src/graph-normalization.ts
export type NormalizedEdgeKind =
  | "relationship"
  | "connection"
  | "satisfy"
  | "verify"
  | "typing"
  | "specializes"
  | "bind"
  | "allocate"
  | "transition"
  | "hierarchy"
  | string;

export interface ArchitectureLikeElement {
  qualified_name?: string | null;
  source_file?: string | null;
}

export function normalizeEdgeKind(relationshipType: string): NormalizedEdgeKind {
  const type = relationshipType.trim().toLowerCase();
  if (!type) return "relationship";
  if (type.includes("item_flow") || type.includes("item flow") || type === "flow" || type.includes("flow")) return "flow";
  if (type.includes("interface-connection") || type.includes("interface connection")) return "interface";
  if (type.includes("interface")) return "interface";
  if (type.includes("binding-connection") || type.includes("binding connection")) return "bind";
  if (type.includes("connection") || type === "connect") return "connection";
  if (type.includes("reference") || type === "ref") return "reference";
  if (type.includes("satisfy")) return "satisfy";
  if (type.includes("verify")) return "verify";
  if (type.includes("derivation") || type.includes("derive")) return "derivation";
  if (type === "typing" || type === "defined_by" || type === "defined by" || type === "definition") return "typing";
  if (type === "dependency" || type.includes("depend") || type.includes("binary-dependency")) return "dependency";
  if (type === "usage" || type === "usage-relationship") return "usage";
  if (type.includes("redefin")) return "redefinition";
  if (type === "specializes" || type === "specialization") return "specializes";
  if (type === "bind" || type === "binding") return "bind";
  if (type === "allocate" || type === "allocation") return "allocate";
  if (type === "transition") return "transition";
  if (type === "composition") return "composition";
  if (type === "hierarchy" || type === "contains" || type === "owns" || type === "ownership" || type === "containment") return "hierarchy";
  return type.replace(/[^a-z0-9_-]+/g, "_") || "relationship";
}

export function isPackageElementType(elementType: string): boolean {
  const normalized = elementType.trim().toLowerCase();
  return (
    !normalized ||
    normalized === "package" ||
    normalized === "library package" ||
    normalized.endsWith("_package") ||
    normalized.includes("package_def")
  );
}

export function isNonDiagramSemanticElementType(elementType: string): boolean {
  const normalized = elementType.trim().toLowerCase();
  if (!normalized) return true;
  // Documentation/comments annotate their owner (SysML §7.4); Spec42 ships them as
  // compartment text only — never as standalone General View boxes ("Unnamed").
  return (
    normalized === "import" ||
    normalized === "documentation" ||
    normalized === "comment" ||
    normalized === "diagnostic" ||
    normalized.includes("diagnostic")
  );
}

export function isOverviewVisualElementType(elementType: string): boolean {
  return !isPackageElementType(elementType) && !isNonDiagramSemanticElementType(elementType);
}

export function isArchitectureElement(element: ArchitectureLikeElement): boolean {
  const qn = element.qualified_name ?? "";
  if (qn.startsWith("Architecture::") || qn.includes("::Architecture::")) {
    return true;
  }
  const source = element.source_file?.trim();
  return source === "Architecture.sysml" || source?.endsWith("/Architecture.sysml") === true;
}

File: crates/diagram_draw/src/graph_normalization.rs
/// Port of `normalizeEdgeKind` in `graph-normalization.ts`.
pub fn normalize_edge_kind(relationship_type: &str) -> String {
    let kind = relationship_type.trim().to_lowercase();
    if kind.is_empty() {
        return "relationship".to_string();
    }
    if kind.contains("item_flow")
        || kind.contains("item flow")
        || kind == "flow"
        || kind.contains("flow")
    {
        return "flow".to_string();
    }
    if kind.contains("interface-connection")
        || kind.contains("interface connection")
        || kind.contains("interface")
    {
        return "interface".to_string();
    }
    if kind.contains("binding-connection")
        || kind.contains("binding connection")
        || kind.contains("binding-connector")
        || kind.contains("binding connector")
    {
        return "bind".to_string();
    }
    if kind.contains("connection") || kind == "connect" || kind.contains("connector") {
        return "connection".to_string();
    }
    if kind.contains("reference") || kind == "ref" {
        return "reference".to_string();
    }
    if kind.contains("satisfy") {
        return "satisfy".to_string();
    }
    if kind.contains("verify") {
        return "verify".to_string();
    }
    if kind.contains("derivation") || kind.contains("derive") {
        return "derivation".to_string();
    }
    if kind == "typing" || kind == "defined_by" || kind == "defined by" || kind == "definition" {
        return "typing".to_string();
    }
    if kind == "dependency" || kind.contains("depend") || kind.contains("binary-dependency") {
        return "dependency".to_string();
    }
    if kind == "usage" || kind == "usage-relationship" {
        return "usage".to_string();
    }
    if kind.contains("redefin") {
        return "redefinition".to_string();
    }
    if kind == "specializes" || kind == "specialization" {
        return "specializes".to_string();
    }
    if kind == "bind" || kind == "binding" {
        return "bind".to_string();
    }
    if kind == "allocate" || kind == "allocation" {
        return "allocate".to_string();
    }
    if kind == "transition" {
        return "transition".to_string();
    }
    if kind == "composition" {
        return "composition".to_string();
    }
    if kind == "hierarchy"
        || kind == "contains"
        || kind == "owns"
        || kind == "ownership"
        || kind == "containment"
    {
        return "hierarchy".to_string();
    }
    // `type.replace(/[^a-z0-9_-]+/g, "_")`: collapse each maximal run of disallowed characters
    // into a single underscore (not per-character, and no trimming of leading/trailing runs).
    let mut sanitized = String::new();
    let mut in_run = false;
    for c in kind.chars() {
        if c.is_ascii_alphanumeric() || c == '_' || c == '-' {
            sanitized.push(c);
            in_run = false;
        } else if !in_run {
            sanitized.push('_');
            in_run = true;
        }
    }
    if sanitized.is_empty() {
        "relationship".to_string()
    } else {
        sanitized
    }
}
